### PR TITLE
WebCard Dynamic Height #870

### PR DESCRIPTION
**Pull request by kevdesilva01**
## Summary

Added dynamic card resizing for the student information card. The card container will adjust in size based on web card content. 

## Changelog

[General] [Add] - Added dynamic web card resizing based on the height of the body of the HTML document. 

## Test Plan

THINGS THAT NEED TO BE IMPLEMENTED FOR THIS TO WORK:
Example implementation: https://cwo-test.ucsd.edu/WebCards/staff_id.html
On any web card HTML file this needs to be added:
```javascript

<script> 
function sendHeight() { 
      var innerHeight = document.getElementById("#id-card").innerHeight(); 
      GetHeight.postMessage(innerHeight); 
} 
</script> 

<body onload="sendHeight()">
```

^^^(this is the initial body tag), to add another function on load just do semicolon and the other function within quotes. (ex: "getData(); sendHeight()" ) 

This implementation does not recognize borders or margins, but it recognizes padding. Other height functions may be considered for this as well. They are commented in the code below. 

- [ ] After the implementation above and the dart implementation on the corresponding card, the web card should adjust its height on the initial load and reload. Max height on card_container.dart may need to be increased as well.

Another implementation is as follows. This contains a bug that causes the card to add an amount to the height (ex: 300 height to 500 height). This causes blank space underneath the content of the web card. It only happens sometimes when scrolling through the home screen and then coming back to the web card. The benefits of this are that it requires no changes on the HTML/js side. The best function that calls for the height must still be decided because each web card may have different circumstances. 

The code is as follows (dart only; as a parameter in the webview widget):
```dart
onPageFinished: (some) async {
  double height = double.parse(
      await _webViewController.evaluateJavascript(
          "document.getElementById('id-card').clientHeight;"));
  //"document.body.clientHeight;")); //"document.documentElement.scrollHeight;"));
  setState(() {
    _contentHeight = height;
    print(height);
  });
},
```

### CODE REVIEW

File: lib/ui/cards/student_info/student_info_card.dart [hunk 1]
```diff
@@ -56,28 +58,39 @@ class _StudentInfoCardState extends State<StudentInfoCard> {
...
-          onWebViewCreated: (controller) {
-            _webViewController = controller;
-          },
-          javascriptChannels: <JavascriptChannel>[
-            _printJavascriptChannel(context),
-          ].toSet(),
-        )),
-      ],
-    );
+    print(url);
```
Comment (i8wu) [inline lib/ui/cards/student_info/student_info_card.dart:61]: remove print

File: lib/ui/cards/student_info/student_info_card.dart [hunk 2]
```diff
@@ -56,28 +58,39 @@ class _StudentInfoCardState extends State<StudentInfoCard> {
...
+            ].toSet(),
+            onPageFinished: (some) async {
+              double height = double.parse(await _webViewController
+                  .evaluateJavascript("document.documentElement.offsetHeight"));
+              setState(() {
+                if (height >= cardContentMinHeight &&
+                    height <= cardContentMaxHeight) {
+                  _contentHeight = double.parse(height.toString());
+                }
+                print("\n\nHeight - " + height.toString());
```
Comment (i8wu) [inline lib/ui/cards/student_info/student_info_card.dart:81]: prints

File: lib/ui/cards/student_info/student_info_card.dart [hunk 3]
```diff
@@ -56,28 +58,39 @@ class _StudentInfoCardState extends State<StudentInfoCard> {
...
+            }));
   }
 
+  //Channel to obtain links and open them in new browser
   JavascriptChannel _printJavascriptChannel(BuildContext context) {
     return JavascriptChannel(
       name: 'CampusMobile',
       onMessageReceived: (JavascriptMessage message) {
         openLink(message.message);
+        print(message.message);
```
Comment (i8wu) [inline lib/ui/cards/student_info/student_info_card.dart:93]: print

File: lib/ui/cards/student_info/student_info_card.dart [hunk 4]
```diff
@@ -56,28 +58,39 @@ class _StudentInfoCardState extends State<StudentInfoCard> {
...
+            javascriptMode: JavascriptMode.unrestricted,
+            initialUrl: url,
+            onWebViewCreated: (controller) {
+              _webViewController = controller;
+            },
+            javascriptChannels: <JavascriptChannel>[
+              _printJavascriptChannel(context),
+            ].toSet(),
+            onPageFinished: (some) async {
+              double height = double.parse(await _webViewController
```
Comment (i8wu) [inline lib/ui/cards/student_info/student_info_card.dart:74]: would recommend moving all this logic to a separate function to keep extra logic out of buildCardContent

File: lib/ui/cards/student_info/student_info_card.dart [hunk 5]
```diff
@@ -56,28 +58,39 @@ class _StudentInfoCardState extends State<StudentInfoCard> {
...
+                  .evaluateJavascript("document.documentElement.offsetHeight"));
+              setState(() {
+                if (height >= cardContentMinHeight &&
+                    height <= cardContentMaxHeight) {
```
Comment (i8wu) [inline lib/ui/cards/student_info/student_info_card.dart:78]: should _contentHeight be set to cardContentMaxHeight if the new height is larger than max? With this current logic the card height will stay min when the height is extra big.

Comment (kevdesilva01) [reply lib/ui/cards/student_info/student_info_card.dart:78]: The original feedback from Charles said to set _contentHeight to the Min height. I can do either one.

**Comment by i8wu** [comment]
The fork does not build.
> lib/core/data_providers/class_schedule_data_provider.dart:4:8: Error: Error when reading 'lib/core/services/class_schedule_service.dart': The system cannot find the file specified.

import 'package:campus_mobile_experimental/core/services/class_schedule_service.dart';
       ^
lib/ui/views/home/home.dart:8:8: Error: Error when reading 'lib/ui/cards/finals/finals_card.dart': The system cannot find the path specified.

import 'package:campus_mobile_experimental/ui/cards/finals/finals_card.dart';
       ^
lib/core/data_providers/class_schedule_data_provider.dart:56:3: Error: Type 'ClassScheduleService' not found.
  ClassScheduleService _classScheduleService;
  ^^^^^^^^^^^^^^^^^^^^
lib/core/data_providers/class_schedule_data_provider.dart:37:29: Error: The method 'ClassScheduleService' isn't defined for the class 'ClassScheduleDataProvider'.
 - 'ClassScheduleDataProvider' is from 'package:campus_mobile_experimental/core/data_providers/class_schedule_data_provider.dart' ('lib/core/data_providers/class_schedule_data_provider.dart').
Try correcting the name to the name of an existing method, or defining a method named 'ClassScheduleService'.
    _classScheduleService = ClassScheduleService();
                            ^^^^^^^^^^^^^^^^^^^^
lib/core/data_providers/class_schedule_data_provider.dart:56:3: Error: 'ClassScheduleService' isn't a type.
  ClassScheduleService _classScheduleService;
  ^^^^^^^^^^^^^^^^^^^^
lib/ui/views/home/home.dart:82:28: Error: The method 'FinalsCard' isn't defined for the class '_HomeState'.
 - '_HomeState' is from 'package:campus_mobile_experimental/ui/views/home/home.dart' ('lib/ui/views/home/home.dart').
Try correcting the name to the name of an existing method, or defining a method named 'FinalsCard'.
          orderedCards.add(FinalsCard());

File: lib/ui/cards/student_info/student_info_card.dart (hunk 5 above)
Comment (i8wu) [reply lib/ui/cards/student_info/student_info_card.dart:78]: I think that's fine for initialize, but I think the behavior I described should be better for taller content. @c3bryant to confirm that behavior.

**Comment by kevdesilva01** [comment]
@i8wu thanks for letting me know, somehow this PR was merged with files from a different PR as this has nothing to do with class_schedule.dart and finals_card.dart. I might have to make a new branch and add all those edits and submit a fresh PR.

**Comment by i8wu** [comment]
@kevdesilva01 No worries. Lots of things are in flight right now so it's difficult to keep track of it all. Thanks for adding all the details to the PR summary, it's definitely helpful for reviewing.

**Pull request closed without merging**